misk-service/src/test/kotlin/misk/CoordinatedServiceTest.kt

```diff
@@ -1,13 +1,46 @@
 package misk
 
+import com.google.common.util.concurrent.AbstractIdleService
 import com.google.common.util.concurrent.Service
+import com.google.inject.Key
 import com.google.inject.Provider
+import com.google.inject.name.Names
 import misk.ServiceGraphBuilderTest.AppendingService
 import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.RepeatedTest
 import org.junit.jupiter.api.Test
+import kotlin.random.Random
 import kotlin.test.assertFailsWith
 
 class CoordinatedServiceTest {
+
+  @RepeatedTest(100) fun fuzzTest() {
```

Comment:
Admittedly, not super happy with the testing strategy here. Was having difficulty reproducing the deadlock in isolation without modifying `CoordinatedService`. Maybe some better ideas around reproduction here? Short of that basically just fuzzing the `ServiceGraphBuilder` - it has caught mistakes, but is definitely not a guarantee of any kind.

Reply:
> Admittedly, not super happy with the testing strategy here.

I'm also not confident that this fuzz test will give use the confidence we need to ensure that we don't end up in a locked state like this again. But trying to detect deadlock is a _very_ hard problem. Testing this is probably always going to be unsatisfying.

Maybe we can leverage a tool like [vmlens](http://vmlens.com/) to help us here? NB: I have not vetted or used vmlens before, it looks like it might be promising.

> Was having difficulty reproducing the deadlock in isolation without modifying CoordinatedService.

Were you able to reproduce this issue reliably in isolation with modifications to CoordinatedService? If so, perhaps we can work out a better testing strategy from there?

@swankjesse might have some ideas.